astropy/io/misc/hdf5.py

```diff
@@ -214,6 +218,45 @@ def _encode_mixins(tbl):
     return encode_tbl
 
 
+def convert_dtype(table, character_as_bytes, copy=False):
+    """
+    Converts bytestring columns (dtype.kind='S') to unicode (dtype.kind='U') and
+    unicode columns (dtype.kind='U') to bytestring (dtype.kind='S').
+
+    Parameters:
+    -----------
+    table: `~astropy.table.Table`
+         Data table in which changes are to be made.
+    character_as_bytes: boolean
+         True if columns are to be converted to bytestrings from unicode.
+         False if columns are to be converted to unicode from bytestrings.
+    copy: boolean
+        True if changes are to be made in a new table which is copied from original table.
+
+    Returns:
+    --------
+    table: `~astropy.table.Table`
+         Data table with desired changes.
+    """
+    if character_as_bytes:
+        colnames = [name for name in table.colnames if table[name].info.dtype.kind == 'U']
+        if colnames and copy:
+            table = table.copy(copy_data=False)
+        try:
+            table.convert_unicode_to_bytestring()
+        except (UnicodeEncodeError, AttributeError):
```

Comment:
Why does this sometimes fail?

Reply:
`convert_unicode_to_bytestring()` converts ASCII-only unicode columns (dtype.kind='U') to bytestring (dtype.kind='S') because of which it throws an error if there are unicode columns present .

Reply:
`convert_unicode_to_bytestring() ` converts ASCII-only unicode columns (dtype.kind='U') to bytestring (dtype.kind='S') because of which it throws an error if there are unicode columns present .